### PR TITLE
main: revise ptags for client tools

### PR DESCRIPTION
**Pull request by masatake**
* [SELF-INCOMPATBILE] emit TAG_KIND_DESCRIPTION ptag only for enabled kinds,
* introduce TAG_FIELD_DESCRIPTION ptag, 
* introduce TAG_EXTRA_DESCRIPTION ptag,
* introduce TAG_PROC_CWD ptag, 
* attach "extras:ptag" fields to ptags, and
* accept ptag names wrapped by '{' and '}'.

TAG_PARAM_DESCRIPTION still reamains in "not-impelemented".
However, the changes in this pull request may be enough for 6.0.

### CODE REVIEW

**Comment by masatake** [comment]
These are should be explaiend in ctags-for-client-developers(5) man page.
We must extend readtags API to support these new ptags.

**Comment by masatake** [comment]
I have written the initial version of ctags-client-tools(7) man page.  See e61022a.

- [ ] Update tags(5)
- [ ] Describe each ptags in ctags-client-tools(7)

I'm also working on readtags side.

With both side changes (ctags and readtags), you can write a linker, a tool merging more than two tags file into one. It is really needed to apply ctags and tags files to large codebases.

The changes in pseudo tags related code are needed to make ctags support multi-pass parsing.

**Comment by masatake** [comment]
I will add TAG_PARAM_DESCRIPTION after releasing 6.0.0.

```
!_TAG_PARAM_DESCRIPTION!C      includePath      /\/usr\/include/
```

File: docs/man/ctags-client-tools.7.rst [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+``--list-pseudo-tags``
+	Lists available kinds of pseudo-tags and shows whether they are enabled or disabled.
+
+	Running ctags with ``--list-pseudo-tags`` option
+	lists available pseudo-tags. Some of pseudo-tags newly introduced
+	in Universal-ctags project are disabled by default. Use
+	``--pseudo-tags=...`` option for enabling them.
+
+``--pseudo-tags=[+|-]names|*``
+	Specifies a list of kinds of pseudo-tag to include in the output.
```
Comment (jwhitley) [inline docs/man/ctags-client-tools.7.rst]:
```suggestion
	A list of kinds of pseudo-tags to include in the output.
```

File: docs/man/ctags-client-tools.7.rst [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+	The parameter names is a set of names listed with
+	``--list-pseudo-tags``. Surround each name of the set by braces
+	like "{TAG_PROGRAM_AUTHOR}". "!_", the prefix representing
+	pseudo-tags is for emitting. You don't have to include the prefix
+	when specifying a name in the option argument for ``--pseudo-tags=``
+	option.
```
Comment (jwhitley) [inline docs/man/ctags-client-tools.7.rst]:
```suggestion
	The parameters are a set of pseudo tag names. Valid pseudo tag names
	can be listed with ``--list-pseudo-tags``. Surround each name in the set 
	with braces, like "{TAG_PROGRAM_AUTHOR}". You don't have to include the "!_" 
	pseudo tag prefix when specifying a name in the option argument for ``--pseudo-tags=``
	option.
```

File: docs/man/ctags-client-tools.7.rst [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+	  $ ctags --extras=+p --pseudo-tags='{TAG_EXTRA_DESCRIPTION}' -o - input.c
+	  !_TAG_EXTRA_DESCRIPTION	anonymous	/Include tags for non-named objects like lambda/
+	  !_TAG_EXTRA_DESCRIPTION	fileScope	/Include tags of file scope/
+	  !_TAG_EXTRA_DESCRIPTION	pseudo	/Include pseudo tags/
+	  !_TAG_EXTRA_DESCRIPTION	subparser	/Include tags generated by subparsers/
+	  ...
+
+	A client tool can know "{anonymous}", "{fileScope}", "{pseudo}",
+	and "{subparser}" extars are enabled from the output.
```
Comment (jwhitley) [inline docs/man/ctags-client-tools.7.rst]:
```suggestion
	and "{subparser}" extras are enabled from the output.
```

File: docs/man/ctags-client-tools.7.rst [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+``TAG_FIELD_DESCRIPTION``  (new in Universal-ctags)
+	Indicates the names and descriptions of enabled fields::
+
+	  !_TAG_FIELD_DESCRIPTION	{field-name}	/description/
+	  !_TAG_FIELD_DESCRIPTION!{language-name}	{field-name}	/description/
+
+	If your tool relies on some fields , refer to the pseudo-tags of
+	this kind.  A tool can reject the tags file that doesn't include
```
Comment (jwhitley) [inline docs/man/ctags-client-tools.7.rst]:
```suggestion
	If your tool relies on some fields, refer to the pseudo-tags of
	this kind.  A tool can reject a tags file that doesn't include
```

File: docs/man/ctags-client-tools.7.rst [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+	TBW
+
+``TAG_OUTPUT_MODE`` (new in Universal-ctags)
+	TBW
+
+``TAG_PATTERN_LENGTH_LIMIT`` (new in Universal-ctags)
+	TBW
+
+``TAG_PROC_CWD`` (new in Universal-ctags)
+	Indicates the working directory of ctags was at.
```
Comment (jwhitley) [inline docs/man/ctags-client-tools.7.rst]:
```suggestion
	Indicates the working directory of ctags during processing.
```

**Comment by masatake** [comment]
Another review. https://github.com/emptyflask/ctags/commit/2f3f88afed69929f88ea31a393b706f76c6ba531from @emptyflask.

I have never imaged I could get more than two reviewers. An issue here; it is hard for me to decide which comments I should choose. I will try to do cherry-pick.

**Comment by AmaiKinono** [comment]
My question after reading the added manpage is: Is it safe to say that tags start with `!_` are pseudo tags? (seems not.) If not, what's the recommended way to deal with pseudo tags for client tools? Seems readtags ignores psuedo tags in its output, what method does it use?

**Comment by masatake** [comment]
We already handled your concerns:-)

ctags can handle "!_" appeared in input well.

input.html
```
<html><body><h1>!_HELLO</h1></body></html>
```

```
$ ctags -o input.tags input.html
$ cat input.tags
!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED	1	/0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_FILESEP	slash	/slash or backslash/
!_TAG_OUTPUT_MODE	u-ctags	/u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR	Universal Ctags Team	//
!_TAG_PROGRAM_NAME	Universal Ctags	/Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL	https://ctags.io/	/official site/
!_TAG_PROGRAM_VERSION	0.0.0	/30783d15/
\x21_HELLO	/tmp/input.html	/^<html><body><h1>!_HELLO<\/h1><\/body><\/html>$/;"	h	language:HTML
[yamato@slave]~/var/ctags-github% readtags -t input.tags  -l 
readtags -t input.tags  -l 
!_HELLO	/tmp/input.html	/^<html><body><h1>!_HELLO<\/h1><\/body><\/html>$/
```

I cannot remember where this behaviour, escaping the first `!`, is documented.
If I cannot find it, we have to wrtite it to our man pages.

For users directly use readtags, the un-escaping feature of readtags is good enough. However, people writing a tool for manipulating tags files may want the way to access raw-escaped data in tags file from the tool. I am the one of the such people.  I'm working on this topic in #2468.

**Comment by masatake** [comment]
I made commits reflecting the reviewing done by @jwhitley and @emptyflask.

**Comment by masatake** [comment]
Though I use "squash!" prefix in the commit logs, I will not squash to allow me to re-read what kind of changes were needed.

**Comment by masatake** [comment]
We should not add TAG_PARAM_DESCRIPTION. If it is needed, individual parsers should emit their own ptags. 

TAG_ROLE_DESCRIPTION, this one is needed when we add an option for enabling/disabling roles.

**Comment by masatake** [comment]
@jwhitley and @emptyflask, thank you very much.
Please, allow me to call you again when I write more documents at #1038.

**Comment by jwhitley** [comment]
I would be glad to help!

**Pull request merged**